### PR TITLE
Updated hdf5: new versions added.

### PR DESCRIPTION
**Pull request by skosukhin**

### CODE REVIEW

**Comment by skosukhin** [comment]
Ok, I don't get it. Why the url that I provide for a particular version must be the same the method url_for_version return?

**Comment by adamjstewart** [comment]
> Ok, I don't get it. Why the url that I provide for a particular version must be the same the method url_for_version returns?

What exactly do you mean? From what I understand, the order of priority is package url, url_for_version, and version url. So if you don't have a version-specific url or url_for_version, it tries to extrapolate from the url. If you have a url_for_version, it uses the value returned from that. And if you have a version-specific url, it always uses that.

Can you update the url_for_version function to return the correct url for this particular version?

**Comment by adamjstewart** [comment]
Oh, I see what you mean. I have no idea why the unit tests are crashing. Can someone take a look at this? There should be no reason that they have to match as far as I know.

**Comment by tgamblin** [comment]
This is actually supposed to be a sanity check for `url_for_version()` to make sure we're extrapolating properly from the per-version URLs.  It's actually useful to make sure the *default* `url_for_version` function is working correctly, but maybe it should be less strict when package authors provide their own implementation.  We could change the test so that it only uses the version in `package.py`.

Given that Spack does consider URLs in the order Adam listed, it *seems* reasonable to have special cases in the package versions that aren't handled by `url_for_version`.  Thoughts?

**Comment by adamjstewart** [comment]
> Given that Spack does consider URLs in the order Adam listed, it seems reasonable to have special cases in the package versions that aren't handled by url_for_version. Thoughts?

Depends on the package. For a one-off version that doesn't match the url for every other version, I'm fine with that. But HDF5 changes its directory layout so often that it's hard to tell whether this is a one-off version or just their new directory layout.

It's intriguing that the version is in a completely different hostname. I wonder if it will be added to the old directory layout soon. This new layout looks like more of a rolling current release download, which will be annoying every time it changes. I'm actually excited for the new release because it fixes problems I was having when trying to compile with NAG.

In summary, I think the version-specific url is the best solution for this version. We just need to fix the unit test first.

**Comment by alalazo** [comment]
@tgamblin Can the logic be moved into the `url_for_version` ? That would solve all the problems and will keep everything in a consistent state, right ?

**Comment by adamjstewart** [comment]
It can be. But the unit test is definitely broken if it disallows such behavior. I think in this case, where it appears that the latest version has it's own "current download" url, it makes more sense to use a version-specific url. That way, when a new version is releases and 1.8.18 is moved to the old layout, the url_for_version won't need to be updated.

But including everything in the url_for_version is definitely prettier...

**Comment by skosukhin** [comment]
So, it works now. I think that url_for_version() method should always give a correct result. That means that it should account for per-version URLs. Maybe later the method should become a single entry point for URL retrievals. That is why I think we should keep the unit test as it is to force package developers to implement their url_for_version() methods properly.
